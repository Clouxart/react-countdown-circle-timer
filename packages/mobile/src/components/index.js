export { CircleTimer } from './CircleTimer'
export { TimeWrapper } from './TimeWrapper'
