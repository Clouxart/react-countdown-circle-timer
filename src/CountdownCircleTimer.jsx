import React, { useEffect, useState, useRef, useMemo } from 'react';
import PropTypes from 'prop-types';
import { useLoop } from '../hooks';
import {
  uuid,
  linearEase,
  getWrapperStyle,
  getTimeStyle,
  svgStyle,
  getPath,
  getNormalizedColors,
  getStroke,
  colorsValidator
} from '../utils';

const getGradientId = (isLinearGradient, gradientUniqueKey) => (
  isLinearGradient ? `countdown-circle-timer-gradient-${gradientUniqueKey || uuid()}` : ''
);

const CountdownCircleTimer = props => {
  const {
    size,
    strokeWidth,
    trailColor,
    durationSeconds,
    isPlaying,
    colors,
    strokeLinecap,
    timeFormatter,
    isLinearGradient,
    gradientUniqueKey
  } = props;

  const pathRef = useRef(null);
  const [pathTotalLength, setPathTotalLength] = useState(0);
  const path = useMemo(() => getPath(size, strokeWidth), [size, strokeWidth]);
  const durationMilliseconds = useMemo(() => durationSeconds * 1000, [durationSeconds]);
  const normalizedColors = useMemo(() => getNormalizedColors(colors, durationMilliseconds, isLinearGradient), [colors, durationMilliseconds, isLinearGradient]);
  const gradientId = useMemo(() => getGradientId(isLinearGradient, gradientUniqueKey), [isLinearGradient, gradientUniqueKey]);

  useEffect(() => {
    const totalLength = pathRef.current.getTotalLength().toFixed(2);
    setPathTotalLength(totalLength);
  }, []);

  const elapsedTime = useLoop(durationMilliseconds, isPlaying);
  const strokeDasharray = linearEase(elapsedTime, 0, pathTotalLength, durationMilliseconds).toFixed(2);
  const stroke = getStroke(normalizedColors, elapsedTime);
  const remainingTime = Math.ceil((durationMilliseconds - elapsedTime) / 1000);

  return (
    <div style={getWrapperStyle(size)}>
      <svg width={size} height={size} style={svgStyle} xmlns="http://www.w3.org/2000/svg">
        {isLinearGradient && (
          <defs>
            <linearGradient id={gradientId} x1="100%" y1="0%" x2="0%" y2="0%">
              {normalizedColors.map(color => <stop {...color.gradient} />)}
            </linearGradient>
          </defs>
        )}
        <path
          fill="none"
          strokeWidth={strokeWidth}
          stroke={trailColor}
          d={path}
        />
        <path
          fill="none"
          stroke={isLinearGradient ? `url(#${gradientId})` : stroke}
          d={path}
          ref={pathRef}
          strokeLinecap={strokeLinecap}
          strokeWidth={strokeWidth}
          strokeDasharray={pathTotalLength}
          strokeDashoffset={strokeDasharray}
        />
      </svg>
      <div style={getTimeStyle(stroke, size)}>
        {typeof timeFormatter === 'function' && timeFormatter(remainingTime, elapsedTime)}
      </div>
    </div>
  );
};

CountdownCircleTimer.propTypes = {
  durationSeconds: PropTypes.number.isRequired,
  colors: PropTypes.arrayOf(
    PropTypes.arrayOf(colorsValidator).isRequired
  ).isRequired,
  size: PropTypes.number,
  strokeWidth: PropTypes.number,
  trailColor: PropTypes.string,
  isPlaying: PropTypes.bool,
  strokeLinecap: PropTypes.oneOf(['round', 'square']),
  timeFormatter: PropTypes.func,
  isLinearGradient: PropTypes.bool,
  gradientUniqueKey: PropTypes.string
};

CountdownCircleTimer.defaultProps = {
  size: 180,
  strokeWidth: 12,
  trailColor: '#d9d9d9',
  isPlaying: false,
  strokeLinecap: 'round',
  isLinearGradient: false
};

export {
  CountdownCircleTimer
};