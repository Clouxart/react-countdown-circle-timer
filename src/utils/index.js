export { uuid } from './uuid'
export { linearEase } from './linearEase'
export { getPathProps } from './getPathProps'
export {
  getWrapperStyle,
  getTimeStyle,
  svgStyle,
  visuallyHidden,
} from './styles'
export { getColorsRGB, getNormalizedColors, getStroke } from './colors'
export { colorsValidator } from './colorsValidator'
