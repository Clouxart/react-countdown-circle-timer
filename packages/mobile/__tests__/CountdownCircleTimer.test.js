import React from 'react'
import { Text } from 'react-native'
import renderer from 'react-test-renderer'

import { CountdownCircleTimer } from '../src'

Math.random = () => 0.124578

const fixture = {
  duration: 10,
  colors: [['#004777', 0.33], ['#F7B801', 0.33], ['#A30000']],
}

describe('snapshot tests', () => {
  it('renders', () => {
    const tree = renderer.create(<CountdownCircleTimer {...fixture} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders with time', () => {
    const tree = renderer
      .create(
        <CountdownCircleTimer
          {...fixture}
          renderAriaTime={({ remainingTime }) => remainingTime}
        >
          {({ remainingTime }) => <Text>{remainingTime}</Text>}
        </CountdownCircleTimer>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
