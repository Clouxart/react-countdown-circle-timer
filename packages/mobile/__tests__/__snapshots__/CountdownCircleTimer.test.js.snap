// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders 1`] = `
<View
  accessibilityLabel="Countdown timer"
  accessible={true}
  style={
    Object {
      "height": 180,
      "position": "relative",
      "width": 180,
    }
  }
>
  <RNSVGSvgView
    bbHeight={180}
    bbWidth={180}
    focusable={false}
    height={180}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 180,
          "width": 180,
        },
      ]
    }
    width={180}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={4292467161}
        strokeWidth={12}
      />
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        stroke={4278208375}
        strokeDasharray={
          Array [
            527.7875658030853,
            527.7875658030853,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={12}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`snapshot tests renders with time 1`] = `
<View
  accessibilityLabel="Countdown timer"
  accessible={true}
  style={
    Object {
      "height": 180,
      "position": "relative",
      "width": 180,
    }
  }
>
  <RNSVGSvgView
    bbHeight={180}
    bbWidth={180}
    focusable={false}
    height={180}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 180,
          "width": 180,
        },
      ]
    }
    width={180}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={4292467161}
        strokeWidth={12}
      />
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        stroke={4278208375}
        strokeDasharray={
          Array [
            527.7875658030853,
            527.7875658030853,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={12}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <Text>
      10
    </Text>
  </View>
  <Text
    accessibilityLiveRegion="assertive"
    accessibilityRole="timer"
    importantForAccessibility="yes"
    style={
      Object {
        "height": 0,
        "opacity": 0,
      }
    }
  >
    10
  </Text>
</View>
`;
